core/domain/expression_parser.py

```diff
@@ -390,11 +412,7 @@ def parse(self, expression):
         # reset to 0.
         self._next_token_index = 0
 
-        parsed_expr = self._parse_expr(token_list)
-        if self._next_token_index < len(token_list):
```

Comment:
Why drop this?

Reply:
This condition is now obsolete since we are inserting the '*' operator between any two operands that have no operator. So, there would be no situation where the parser is trying to look for an operator and it can't find it. Previously, an expression like `a(b+c)` would've been considered as invalid but since we would now change it to `a*(b+c)`, it will be considered as a valid expression.

Reply:
Aren't there other ways to be invalid -- e.g. lots of consecutive operators, maybe an input string like "-----+"?

Reply:
Nope. Everything else is covered in their respective places. For instance, the eg that you gave is being caught in the `_parse_pow_expr` method where the production rule of the grammar is supposed to handle the + and - unary operators.
https://github.com/oppia/oppia/blob/ae8f1426d70ab82702502b7e8461a4f7cb5b2afc/core/domain/expression_parser.py#L448